### ISSUE TITLE
pyproj-2.3.0 segfault in test_datum on Windows wheels

### ISSUE THREAD

**Issue by cgohlke** [pyproj4/pyproj#403]
#### Code Sample, a copy-pastable example if possible

Copied from [test_datum.py](https://github.com/pyproj4/pyproj/blob/master/test/test_datum.py):
```python
import faulthandler;faulthandler.enable()
from numpy.testing import assert_almost_equal
from pyproj import Proj, transform
p1 = Proj(proj="latlong", datum="WGS84")
s_1 = -111.5
s_2 = 45.25919444444
p2 = Proj(proj="utm", zone=10, datum="NAD27")
x2, y2 = transform(p1, p2, s_1, s_2)
```
#### Problem description

The above code crashes on any Python 3 version due to access violation:
```
Windows fatal exception: access violation

Current thread 0x000044cc (most recent call first):
  File "X:\Python38\lib\site-packages\pyproj\transformer.py", line 274 in from_crs
  File "X:\Python38\lib\site-packages\pyproj\transformer.py", line 237 in from_proj
  File "X:\Python38\lib\site-packages\pyproj\transformer.py", line 674 in transform
  File "<stdin>", line 1 in <module>
```

The crash is at https://github.com/pyproj4/pyproj/blob/11b769cc3b808a7f417103d0b033222b07737cec/pyproj/_transformer.pyx#L252

#### Expected Output

No crash.  pyproj-2.2.2 does not crash.

#### Environment Information

 - Output from: `python -c "import pyproj; pyproj.show_versions()"`
```
System:
    python: 3.8.0b3 (tags/v3.8.0b3:4336222, Jul 29 2019, 16:54:03) [MSC v.1916 64 bit (AMD64)]
executable: X:\Python38\python.exe
   machine: Windows-10-10.0.18362-SP0

PROJ:
      PROJ: 6.1.1
  data dir: X:\Python38\lib\site-packages\pyproj\proj_dir\share\proj

Python deps:
    pyproj: 2.3.0
       pip: 19.2.2
setuptools: 41.1.0
    Cython: 0.29.13
```

#### Installation method

Built from source against proj-6.1.1, sqlite-amalgamation-3290000.

**Comment by snowman2** [comment]
Thanks for the report! :thinking:

**Comment by snowman2** [comment]
Do all of the other tests work okay or do others segfault as well?

**Comment by snowman2** [comment]
I wonder why it does not segfault here: https://ci.appveyor.com/project/jswhit/pyproj-coed1/builds/26786573

**Comment by snowman2** [comment]
Do you have PROJ 5.2.0 installed on your machine by chance? https://github.com/pyproj4/pyproj/issues/368

Unlikely since it worked with 2.2.2, but usually better not to assume :smile:

**Comment by cgohlke** [comment]
Several other test segfault, including
```
test/test_datum_shift.py::test_shift_wgs84_to_gaussb_no_ellisoidal_height
test/test_datum_shift.py::test_shift_gaussb_to_wgs84_no_ellisoidal_height
test/test_doctest_wrapper.py::test_doctests
```

I'll start over with new copies of source code...

**Comment by cgohlke** [comment]
> I'll start over with new copies of source code...

No luck. The crash is in the `sqlite3` library. Maybe related to the sqlite build flags, but why wouldn't pyproj-2.2.2 crash with the same dependencies?

```
 	0000000000000000()	Unknown
 	_transformer.cp38-win_amd64.pyd!sqlite3_malloc()	C
 	_transformer.cp38-win_amd64.pyd!sqlite3_str_vappendf()	C
 	_transformer.cp38-win_amd64.pyd!sqlite3_prepare()	C
 	_transformer.cp38-win_amd64.pyd!sqlite3_prepare()	C
 	_transformer.cp38-win_amd64.pyd!sqlite3_prepare_v2()	C
 	_transformer.cp38-win_amd64.pyd!osgeo::proj::io::DatabaseContext::Private::run(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::list<struct osgeo::proj::io::SQLValues,class std::allocator<struct osgeo::proj::io::SQLValues> > const &)	C++
 	_transformer.cp38-win_amd64.pyd!osgeo::proj::io::AuthorityFactory::createObjectsFromName(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::vector<enum osgeo::proj::io::AuthorityFactory::ObjectType,class std::allocator<enum osgeo::proj::io::AuthorityFactory::ObjectType> > const &,bool,unsigned __int64)	C++
 	_transformer.cp38-win_amd64.pyd!std::_Tree_const_iterator<class std::_Tree_val<struct std::_Tree_simple_types<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > > > >::operator--(void)	C++
 	_transformer.cp38-win_amd64.pyd!osgeo::proj::operation::ParameterValue::booleanValue(void)	C++
 	_transformer.cp38-win_amd64.pyd!std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >::find(char const * const,unsigned __int64)	C++
 	_transformer.cp38-win_amd64.pyd!osgeo::proj::operation::CoordinateOperationFactory::Private::createOperations(class dropbox::oxygen::nn<class std::shared_ptr<class osgeo::proj::crs::CRS> > const &,class dropbox::oxygen::nn<class std::shared_ptr<class osgeo::proj::crs::CRS> > const &,struct osgeo::proj::operation::CoordinateOperationFactory::Private::Context &)	C++
 	_transformer.cp38-win_amd64.pyd!osgeo::proj::operation::CoordinateOperationFactory::createOperations(class dropbox::oxygen::nn<class std::shared_ptr<class osgeo::proj::crs::CRS> > const &,class dropbox::oxygen::nn<class std::shared_ptr<class osgeo::proj::crs::CRS> > const &,class dropbox::oxygen::nn<class std::unique_ptr<class osgeo::proj::operation::CoordinateOperationContext,struct std::default_delete<class osgeo::proj::operation::CoordinateOperationContext> > > const &)	C++
 	_transformer.cp38-win_amd64.pyd!proj_create_operations()	C++
 	_transformer.cp38-win_amd64.pyd!proj_create_crs_to_crs()	C++
>	_transformer.cp38-win_amd64.pyd!__pyx_pf_6pyproj_12_transformer_12_Transformer_6from_crs(__pyx_obj_6pyproj_4_crs__CRS * __pyx_v_crs_from=0x000001a585dd7e10, __pyx_obj_6pyproj_4_crs__CRS * __pyx_v_crs_to=0x000001a585dd76c0, _object * __pyx_v_skip_equivalent=0x00007ffd5e040760, _object * __pyx_v_always_xy=0x00007ffd5e040760, _object * __pyx_v_area_of_interest=0x00007ffd5e04cb30) Line 4438	C
 	_transformer.cp38-win_amd64.pyd!__pyx_pw_6pyproj_12_transformer_12_Transformer_7from_crs(_object * __pyx_self=0x0000000000000000, _object * __pyx_args=0x000001a585ad1e40, _object * __pyx_kwds=0x000001a585af8d40) Line 4161	C
 	python38.dll!cfunction_call_varargs(_object * func=0x000001a5857872c0, _object * args=0x000001a585ad1e40, _object * kwargs=0x000001a585af8d40) Line 742	C
 	[Inline Frame] python38.dll!PyCFunction_Call(_object *) Line 772	C
 	[Inline Frame] python38.dll!_PyObject_MakeTpCall(_object * keywords=0x000001a58523cb40, _object * const *) Line 159	C
 	[Inline Frame] python38.dll!_PyObject_Vectorcall(_object * nargsf, _object * const *) Line 125	C
...
```

**Comment by cgohlke** [comment]
pyproj-2.2.2 (built with same libraries and compilers) also crashes, but not every time and at a different point:
```
test/test_crs.py::test_datum Windows fatal exception: access violation
```
```
 	_crs.cp38-win_amd64.pyd!pj_ctx_get_errno()	C++
 	_crs.cp38-win_amd64.pyd!pj_free()	C++
 	_crs.cp38-win_amd64.pyd!proj_destroy()	C++
>	_crs.cp38-win_amd64.pyd!__pyx_pf_6pyproj_4_crs_4Base_2__dealloc__(__pyx_obj_6pyproj_4_crs_Base * __pyx_v_self=0x00000165d26e6a40) Line 5579	C
 	_crs.cp38-win_amd64.pyd!__pyx_pw_6pyproj_4_crs_4Base_3__dealloc__(_object * __pyx_v_self=0x00000165d26e6a40) Line 5537	C
 	_crs.cp38-win_amd64.pyd!__pyx_tp_dealloc_6pyproj_4_crs_Base(_object * o=0x00000165d26e6a40) Line 20542	C
 	_crs.cp38-win_amd64.pyd!__pyx_tp_dealloc_6pyproj_4_crs_Datum(_object * o=0x00000165d26e6a40) Line 21014	C
 	python38.dll!frame_dealloc(_frame * f=0x00000165b4326da0) Line 417	C
```

**Comment by snowman2** [comment]
Hmmm, interesting. Did you just upgrade your compiler?

**Comment by snowman2** [comment]
On appveyor, we are using:
```
      VS_VERSION: Visual Studio 14
      APPVEYOR_BUILD_WORKER_IMAGE: Visual Studio 2015
```

**Comment by cgohlke** [comment]
> Did you just upgrade your compiler?

I can reproduce these crashes with Visual Studio 2017 (v15.9.15) and 2015 (v14 Update 3), any recent Python 3 version, 32-bit and 64-bit. Also tried downgrading to sqlite-3250100 to no avail.

**Comment by snowman2** [comment]
I wonder if it is related to my fix for #374

**Comment by cgohlke** [comment]
Could be. I am linking statically, which worked so far.

**Comment by cgohlke** [comment]
Linking against the proj DLL fixes the segfaults so there is probably an issue sharing a single PJ_CONTEXT across Cython extensions when linking statically .

**Comment by snowman2** [comment]
This is unfortunate. Will probably not be able to do anything until after https://github.com/OSGeo/PROJ/pull/1566

**Comment by snowman2** [comment]
This should be addressed in 2.4.0
